### PR TITLE
dev-lang/zig: new package

### PR DESCRIPTION
**Pull request by nrdmn**
Zig is an open-source programming language designed for robustness, optimality, and maintainability.

https://github.com/ziglang/zig
Gentoo-Bug: https://bugs.gentoo.org/683386

### CODE REVIEW

**Comment by nrdmn** [comment]
ping

File: dev-lang/zig/metadata.xml [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+	</maintainer>
+	<maintainer type="project">
+		<email>proxy-maint@gentoo.org</email>
+		<name>Proxy Maintainers</name>
+	</maintainer>
+	<upstream>
+		<remote-id type="github">ziglang/zig</remote-id>
+	</upstream>
+	<longdescription lang="en">
+		Zig is a general-purpose programming language designed for robustness, optimality, and maintainability.
```
Comment (juippis) [inline dev-lang/zig/metadata.xml]: This isn't really any longer than the default DESCRIPTION in ebuild.

File: dev-lang/zig/zig-0.4.0.ebuild [hunk 2]
```diff
@@ -0,0 +1,50 @@
+# Copyright 2019 Gentoo Authors
+# Distributed under the terms of the GNU General Public License v2
+
+EAPI=7
```
Comment (juippis) [inline dev-lang/zig/zig-0.4.0.ebuild]: One empty line below EAPI please, to follow default skel style.

File: dev-lang/zig/zig-0.4.0.ebuild [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+inherit cmake-utils
+
+DESCRIPTION="A robust, optimal, and maintainable programming language."
```
Comment (juippis) [inline dev-lang/zig/zig-0.4.0.ebuild]: No dot here in the end please :)

File: dev-lang/zig/zig-0.4.0.ebuild [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+HOMEPAGE="https://ziglang.org/"
+LICENSE="MIT"
+SLOT="0"
+IUSE=""
+
+if [[ ${PV} == 9999 ]]; then
+	EGIT_REPO_URI="https://github.com/ziglang/zig.git"
+	inherit git-r3
+else
+	SRC_URI="https://github.com/ziglang/zig/archive/${PV}.tar.gz -> zig-${PV}.tar.gz"
```
Comment (juippis) [inline dev-lang/zig/zig-0.4.0.ebuild]: Why not `-> ${P}.tar.gz`?

File: dev-lang/zig/zig-0.4.0.ebuild [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+	KEYWORDS="~amd64 ~x86"
```
Comment (juippis) [inline dev-lang/zig/zig-0.4.0.ebuild]: Did you test this on x86? Asking due to complicated llvm dep.

File: dev-lang/zig/zig-0.4.0.ebuild [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+fi
+
+ALL_LLVM_TARGETS=( AArch64 AMDGPU ARM BPF Hexagon Lanai Mips MSP430 NVPTX
+	PowerPC Sparc SystemZ WebAssembly X86 XCore )
+ALL_LLVM_TARGETS=( "${ALL_LLVM_TARGETS[@]/#/llvm_targets_}" )
+# According to zig's author, zig builds that do not support all targets are not
+# supported by the upstream project.
+LLVM_TARGET_USEDEPS=${ALL_LLVM_TARGETS[@]}
```
Comment (juippis) [inline dev-lang/zig/zig-0.4.0.ebuild]: Have you tried and confirmed this? This is really scary and to me doesnt make any sense enabling all targets. I'd personally add a some sort of 'experimental' USE flag where, if it works, you could run this without everything enabled. Then just print some info message that upstream support, and maybe support altogether, isnt offered due to experimental flag. 

I dont have answers to this, just looks a bit.. suspicious and scary :)

File: dev-lang/zig/zig-0.4.0.ebuild [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+
+RDEPEND="~sys-devel/llvm-8.0.0:=[${LLVM_TARGET_USEDEPS// /,}]"
```
Comment (juippis) [inline dev-lang/zig/zig-0.4.0.ebuild]: I think the subslot binding is unnecessary here, portage should prevent changing LLVM_TARGETS on its own, I guess that was the purpose of it?

File: dev-lang/zig/zig-0.4.0.ebuild [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+DEPEND="
+	${RDEPEND}
+	dev-util/cmake
```
Comment (juippis) [inline dev-lang/zig/zig-0.4.0.ebuild]: cmake is already depended upon from cmake-utils.eclass. It's also a BDEPEND ;)

File: dev-lang/zig/zig-0.4.0.ebuild [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+	|| (
```
Comment (juippis) [inline dev-lang/zig/zig-0.4.0.ebuild]: This isn't the correct way to check if gcc or clang requirements are satisfied. I wouldnt worry about this honestly, as these versions are so old default profiles should already depend on newer ones. I dont think there even is clang-3 in the tree anymore. 

However if you truly wish to make these checks, use toolchain-funcs.eclass and pkg_setup phase instead.

File: dev-lang/zig/zig-0.4.0.ebuild [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+		>=sys-devel/gcc-5.0.0
+		>=sys-devel/clang-3.6.0
+	)
+"
+
+src_prepare() {
+	# Suppress error messages
+	sed -i '/if(NOT(CMAKE_BUILD_TYPE/,/endif()/d' cmake/Findllvm.cmake
```
Comment (juippis) [inline dev-lang/zig/zig-0.4.0.ebuild]: External commands need to be appended with `|| die "message what went wrong"`.

These comments apply to the -9999 ebuild too.

File: dev-lang/zig/zig-0.4.0.ebuild (hunk 6 above)
Comment (nrdmn) [reply dev-lang/zig/zig-0.4.0.ebuild]: If zig is used to compile for an architecture that the installed llvm does not support, it gives a misleading error message:
```
$ zig build-exe -target powerpc-linux-gnu hello.zig
Zig is expecting LLVM to understand this target: 'powerpc-unknown-linux-gnu'
However LLVM responded with: "No available targets are compatible with triple "powerpc-unknown-linux-gnu""
Zig is unable to continue. This is a bug in Zig:
https://github.com/ziglang/zig/issues/438
```
I've added an experimental use flag as you suggested.

File: dev-lang/zig/zig-0.4.0.ebuild (hunk 5 above)
Comment (nrdmn) [reply dev-lang/zig/zig-0.4.0.ebuild]: No, i figured it would work on x86 if it works on amd64

**Comment by juippis** [comment]
Looks good :) something I got while installing:
```
 * The ebuild is installing to one or more unexpected paths:
 * 
 *   /tmp
 * 
 * Please fix the ebuild to use correct FHS/Gentoo policy paths.
```
```
# equery f zig | grep /tmp/
/tmp/portage
/tmp/portage/dev-lang
/tmp/portage/dev-lang/zig-0.4.0
/tmp/portage/dev-lang/zig-0.4.0/work
/tmp/portage/dev-lang/zig-0.4.0/work/zig-0.4.0_build
/tmp/portage/dev-lang/zig-0.4.0/work/zig-0.4.0_build/zig_cpp
/tmp/portage/dev-lang/zig-0.4.0/work/zig-0.4.0_build/zig_cpp/libembedded_lld_coff.a
/tmp/portage/dev-lang/zig-0.4.0/work/zig-0.4.0_build/zig_cpp/libembedded_lld_elf.a
/tmp/portage/dev-lang/zig-0.4.0/work/zig-0.4.0_build/zig_cpp/libembedded_lld_lib.a
/tmp/portage/dev-lang/zig-0.4.0/work/zig-0.4.0_build/zig_cpp/libembedded_lld_mingw.a
/tmp/portage/dev-lang/zig-0.4.0/work/zig-0.4.0_build/zig_cpp/libembedded_lld_wasm.a
/tmp/portage/dev-lang/zig-0.4.0/work/zig-0.4.0_build/zig_cpp/libzig_cpp.a
```

That needs to be fixed.

I tried some examples from its homepage and it _seems_ to work fine with experimental USE, 
```
[ebuild   R    ] sys-devel/llvm-8.0.0:8::gentoo  USE="libffi ncurses xml -debug -doc -exegesis -gold -libedit -test -xar" ABI_X86="32 (64) (-x32)" LLVM_TARGETS="AMDGPU BPF (X86) -AArch64 -ARM -Hexagon -Lanai -MSP430 -Mips -NVPTX -PowerPC -Sparc -SystemZ -WebAssembly -XCore" 0 KiB
```

**Comment by nrdmn** [comment]
Good catch! I'm not really sure what to do with these files as they're not included in upstream's builds. They are not being built when using the master branch. I removed them from CMakeLists.txt.

**Comment by nrdmn** [comment]
There was a change on zig's master branch a few hours ago that breaks the build of zig-9999.ebuild. I'll push an updated zig-9999.ebuild later today.

**Comment by juippis** [comment]
Can't test this yet, but I noticed the tree has llvm-8.0.1, is zig supposed to work with that? A legit question, not a complaint :)

**Comment by nrdmn** [comment]
It seems to work with llvm-8.0.1. I've updated the ebuilds. Thanks for your feedback!

**Comment by juippis** [comment]
```
  dependency.missingslot        2
   dev-lang/zig/zig-0.4.0.ebuild: RDEPEND: 'sys-devel/llvm[llvm_targets_AArch64,llvm_targets_AMDGPU,llvm_targets_ARM,llvm_targets_BPF,llvm_targets_Hexagon,llvm_targets_Lanai,llvm_targets_Mips,llvm_targets_MSP430,llvm_targets_NVPTX,llvm_targets_PowerPC,llvm_targets_Sparc,llvm_targets_SystemZ,llvm_targets_WebAssembly,llvm_targets_X86,llvm_targets_XCore]' matches more than one slot, please specify an explicit slot and/or use the := or :* slot operator
   dev-lang/zig/zig-9999.ebuild: RDEPEND: 'sys-devel/llvm[llvm_targets_AArch64,llvm_targets_AMDGPU,llvm_targets_ARM,llvm_targets_BPF,llvm_targets_Hexagon,llvm_targets_Lanai,llvm_targets_Mips,llvm_targets_MSP430,llvm_targets_NVPTX,llvm_targets_PowerPC,llvm_targets_Sparc,llvm_targets_SystemZ,llvm_targets_WebAssembly,llvm_targets_X86,llvm_targets_XCore]' matches more than one slot, please specify an explicit slot and/or use the := or :* slot operator
  ebuild.minorsyn               3
   dev-lang/zig/zig-9999.ebuild: Unquoted Variable on line: 49
   dev-lang/zig/zig-9999.ebuild: Unquoted Variable on line: 56
   dev-lang/zig/zig-9999.ebuild: Unquoted Variable on line: 60
```
;)

This is just about ready to be merged. You can see some common mistakes by running `repoman full -dx` in zig's ebuild directory.

File: dev-lang/zig/zig-0.4.0.ebuild [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+ALL_LLVM_TARGETS=( AArch64 AMDGPU ARM BPF Hexagon Lanai Mips MSP430 NVPTX
+	PowerPC Sparc SystemZ WebAssembly X86 XCore )
+ALL_LLVM_TARGETS=( "${ALL_LLVM_TARGETS[@]/#/llvm_targets_}" )
+# According to zig's author, zig builds that do not support all targets are not
+# supported by the upstream project.
+LLVM_TARGET_USEDEPS=${ALL_LLVM_TARGETS[@]}
+
+RDEPEND="
+	=sys-devel/llvm-8*
+	!experimental? ( sys-devel/llvm[${LLVM_TARGET_USEDEPS// /,}] )
```
Comment (juippis) [inline dev-lang/zig/zig-0.4.0.ebuild]: This is still missing the slot and repoman yells about it. I believe the syntax used above works here too, but remember to run repoman to test it.

File: dev-lang/zig/zig-9999.ebuild [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+	local mycmakeargs=(
+		-DBUILD_SHARED_LIBS=OFF
+		-DDESTDIR="${T}"
+	)
+
+	cmake-utils_src_configure
+}
+
+src_install() {
+	cp -r "${T}/usr" "${ED}"
```
Comment (juippis) [inline dev-lang/zig/zig-9999.ebuild]: This is an external command and missing `|| die "message"`

File: dev-lang/zig/zig-9999.ebuild [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+
+	cmake-utils_src_install
+
+	rm -rf "${ED}/var"
```
Comment (juippis) [inline dev-lang/zig/zig-9999.ebuild]: Same here (about die), although -f makes it so it never dies even if it doesnt do anything. Wont it work without?

This should be last miiiiinor mistakes ;)

**Comment by juippis** [comment]
```
>>> Working in BUILD_DIR: "/tmp/portage/dev-lang/zig-9999/work/zig-9999_build"
[0/3] cd /tmp/portage/dev-lang/zig-9999/work/zig-9999 && /tmp/portage/dev-lang/zig-9999/work/zig-9999_build/zig0 build --override-std-dir std --override-lib-dir /tmp/portage/dev-lang/zig-9999/work/zig-9999 libuserland install -Doutput-dir=/tmp/portage/dev-lang/zig-9999/work/zig-9999_build -Dlib-files-only --[1/3] cd /tmp/portage/dev-lang/zig-9999/work/zig-9999 && /tmp/portage/dev-lang/zig-9999/work/zig-9999_build/zig0 build --override-std-dir std --override-lib-dir /tmp/portage/dev-lang/zig-9999/work/zig-9999 libuserland install -Doutput-dir=/tmp/portage/dev-lang/zig-9999/work/zig-9999_build -Dlib-files-only --[1/2] Install the project...
-- Install configuration: "Gentoo"
-- Installing: /tmp/portage/dev-lang/zig-9999/image/usr/bin/zig
-- Set runtime path of "/tmp/portage/dev-lang/zig-9999/image/usr/bin/zig" to ""
rm: cannot remove '/tmp/portage/dev-lang/zig-9999/image/var': No such file or directory
 * ERROR: dev-lang/zig-9999::testworld failed (install phase):
 *   unable to remove /var
 * 
 * Call stack:
 *     ebuild.sh, line 125:  Called src_install
 *   environment, line 2636:  Called die
 * The specific snippet of code:
 *       rm -r "${ED}/var" || die "unable to remove /var"
```

The 0.4.0 works. I can commit it if you want to fix -9999 another day. 

Then I realized something: To make user experience seamless, experimental flag should probably be enabled by default. We got this the wrong way around :\

So either do `IUSE="+experimental"` or flip it so enabling a USE flag requires llvm with all its deps.
Also, do this: ` =sys-devel/llvm-8*[${LLVM_TARGET_USEDEPS// /,}]`, not `:=` ;)

**Comment by nrdmn** [comment]
That's strange, both ebuilds work fine for me. But I've put the `f` back where it was.

**Comment by juippis** [comment]
Hey, I decided not to include the -9999 ebuild. It still doesn't work for me, and the -f isn't really a solution. You need to know when it's present and when it's not. Could the state of use flags make a difference? 

Let's merge 0.4.0 and close this since it's good and works. Please open a new PR to add a live ebuild, and lets work it out there more in detail. FYI, it installs TONS of files into /tmp which is not acceptable.

**Comment by Zlogene** [comment]
why was not llvm subslotted properly? it has slot :8 to properly depend on it

**Comment by juippis** [comment]
Because I thought `=sys-devel/llvm-8*` is enough to handle and bind the dependency. I can add the slot and test it in a minute.

**Comment by Zlogene** [comment]
`=` binds to 8.x and `:8` does the same, although this is better avvoiding equal sign in denendencies if possible

**Comment by juippis** [comment]
efbc7eb4725347d59e62e66403b037b17d4223b0
I tested whether llvm.eclass was needed for this. @Zlogene please confirm whether this is ok?

**Comment by Zlogene** [comment]
looks good @juippis !

**Pull request closed without merging**